### PR TITLE
refactor: improve lifecycle observer and session handler architecture

### PR DESCRIPTION
**Pull request by SKannaniOS**
## Description
This PR refactors the lifecycle management and session handling architecture to improve code maintainability, thread safety, and separation of concerns. The main changes include:

- **Architectural Improvement**: Replaced the `LifecycleSessionWrapper` with direct management of `LifecycleObserver` and `SessionHandler` components within the `Analytics` class
- **Thread Safety Enhancement**: Added `@Synchronized` property wrapper to the observers array in `LifecycleObserver` to prevent race conditions
- **Memory Management**: Improved cleanup mechanisms and reference management to prevent memory leaks
- **Session Initialization**: Enhanced session handler initialization to automatically start sessions and attach observers based on configuration
- **Code Simplification**: Removed redundant wrapper class and simplified component interactions
- **Test Coverage**: Added comprehensive test coverage with `MockLifecycleEventListener` and improved existing tests

## Type of change
- [x] Code refactor/optimization
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### Core Changes
- **Removed `LifecycleSessionWrapper`**: Eliminated the wrapper class and moved lifecycle observer and session handler management directly into the `Analytics` class
- **Enhanced `LifecycleObserver`**: 
  - Removed dependency on `Analytics` instance in constructor
  - Added thread-safe observer management with `@Synchronized` property wrapper
  - Improved event handling with snapshot-based observer notification to prevent deadlocks
- **Improved `SessionHandler`**: 
  - Added automatic session initialization in constructor based on configuration
  - Enhanced observer attachment/detachment lifecycle management
  - Added `isSessionManual` property for cleaner session type checking
- **Updated `Analytics`**: 
  - Direct management of `lifecycleObserver` and `sessionHandler` properties  
  - Improved cleanup in `shutdownHook()` with dedicated `referenceCleanup()` method
  - Enhanced component initialization order

### Test Improvements
- **New Mock**: Added `MockLifecycleEventListener` with comprehensive testing utilities
- **Enhanced Test Coverage**: Improved test coverage for lifecycle events, session management, and edge cases
- **Async Testing**: Added proper async/await testing patterns for lifecycle event handling
- **Edge Case Testing**: Added tests for system restart scenarios, session timeouts, and configuration changes

## Checklist
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added the necessary documentation (if appropriate).
- [x] I have ensured that my code follows the project's code style.
- [x] I have checked for potential performance impacts and optimized if necessary.
- [x] I have checked the code for security issues.
- [ ] I have updated the changelog (if required).

### Automated Testing
1. Run the existing test suite to ensure no regressions:
   ```bash
   xcodebuild test -scheme RudderStackAnalytics -destination 'platform=iOS Simulator,name=iPhone 15'
   ```

2. The new tests specifically cover:
   - Lifecycle observer thread safety with multiple observers
   - Session handler initialization with different configurations
   - Proper cleanup and memory management
   - Edge cases like system restarts and session timeouts

### Manual Testing
1. **Session Management**: Test automatic and manual session behaviors
2. **Lifecycle Events**: Verify app foreground/background transitions work correctly
3. **Memory Management**: Test that observers are properly cleaned up on deallocation
4. **Configuration Changes**: Test behavior with different session configuration settings

## Breaking Changes
This refactoring maintains API compatibility. No public API changes have been made, so existing integrations should continue to work without modification.

## Maintainers Checklist
- [ ] The code has been reviewed.
- [ ] CI tests have passed.
- [ ] All necessary documentation has been updated.

## Screenshots (if applicable)
N/A - Internal architecture refactoring with no UI changes.

## Additional Context
This refactoring addresses several architectural concerns:

1. **Separation of Concerns**: The previous `LifecycleSessionWrapper` was an unnecessary abstraction that coupled lifecycle observation with session management. This change provides cleaner separation.

2. **Thread Safety**: The `@Synchronized` wrapper ensures thread-safe access to the observers array, preventing potential race conditions in multi-threaded scenarios.

3. **Memory Management**: Improved cleanup mechanisms reduce the risk of memory leaks and ensure proper resource deallocation.

4. **Testing**: The addition of `MockLifecycleEventListener` and enhanced test coverage provides better confidence in the system's reliability.

5. **Maintainability**: Removing the wrapper class and simplifying the architecture makes the codebase easier to understand and maintain.

The changes maintain full backward compatibility while providing a more robust and maintainable foundation for future development.

### CODE REVIEW

File: Sources/RudderStackAnalytics/SessionManagement/SessionHandler.swift [hunk 1]
```diff
@@ -58,30 +61,50 @@ final class SessionHandler {
...
+            } else if !shouldBeRegistered && registered {
+                self.analytics.lifecycleObserver?.removeObserver(self)
+                registered = false
+            }
+        }
     }
     
-    func deregisterObserver() {
+    /// Force deregisters the observer without state checks.
+    func forceDeregisterObserver() {
```
Comment (1abhishekpandey) [inline Sources/RudderStackAnalytics/SessionManagement/SessionHandler.swift]: Let's have it as deregisterObserver only. There's no need to use `force` word.

File: Sources/RudderStackAnalytics/SessionManagement/SessionHandler.swift [hunk 2]
```diff
@@ -23,6 +23,9 @@ final class SessionHandler {
     private var storage: KeyValueStorage
     private var sessionState: StateImpl<SessionInfo>
     
+    /// Tracks current observer registration state to avoid redundant operations
+    @Synchronized private var isObserverRegistered: Bool = false
```
Comment (1abhishekpandey) [inline Sources/RudderStackAnalytics/SessionManagement/SessionHandler.swift]: Can you try to improve the code such that you don't require any state variable to add an observer? 
(Optional: Refer to the Kotlin code if required.)


```suggestion
```

File: Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift [hunk 3]
```diff
@@ -63,6 +63,7 @@ extension LifecycleObserver {
 // MARK: - Observer Management
 extension LifecycleObserver {
     func addObserver(_ observer: LifecycleEventListener) {
+        guard !observers.contains(where: { $0.observer === observer }) else { return }
```
Comment (1abhishekpandey) [inline Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift]: `addObserver` should only be responsible for adding and not checking if this observer is already added or not and then taking the action. This is a violation of SRP.

Let's ensure we call this function only when it's needed, without the need to check anything.

```suggestion
```

Comment (1abhishekpandey) [reply Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift]: Ensure we've full parity with Kotlin, as that has the correct implementation. By fixing the actual issue, i.e., ensure we call the `addObserver` and `removeObserver` only once in the session class, as Kotlin does.

This would simplify the code.

Comment (1abhishekpandey) [reply Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift]: Also, I noticed this just now: Make this [observers](https://github.com/rudderlabs/rudder-sdk-swift/blob/c7f2c8cd0cd9a98fd53145cb3098aac444212122/Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift#L17) thread safe.

**Review by 1abhishekpandey — CHANGES_REQUESTED** [review]
Let's also address the AI comments.

File: Sources/RudderStackAnalytics/SessionManagement/SessionHandler.swift (hunk 2 above)
Comment (1abhishekpandey) [reply Sources/RudderStackAnalytics/SessionManagement/SessionHandler.swift]: Let's continue the discussion over [here](https://github.com/rudderlabs/rudder-sdk-swift/pull/199#discussion_r2675139723).

File: Sources/RudderStackAnalytics/SessionManagement/SessionHandler.swift [hunk 4]
```diff
@@ -33,55 +32,59 @@ final class SessionHandler {
...
+            self.attachSessionTrackingObservers()
+        } else if !isSessionManual {
+            self.endSession()
+        }
+    }
+    
+    private func checkAndStartSessionOnLaunch() {
+        guard self.sessionId == nil || self.isSessionManual || self.isSessionTimedOut else { return }
+        self.startSession(id: Self.generatedSessionId, type: .automatic)
     }
```
[the remark this answers is not in the document]
Comment (SKannaniOS) [reply Sources/RudderStackAnalytics/SessionManagement/SessionHandler.swift:47]: This is the expected behavior.

File: Sources/RudderStackAnalytics/SessionManagement/SessionHandler.swift [hunk 5]
```diff
@@ -33,55 +32,59 @@ final class SessionHandler {
...
     
     func startSession(id: UInt64, type: SessionType) {
         self.updateSessionStart(isSessionStart: true)
         self.updateSessionType(type: type)
+        if isSessionManual {
+            detachSessionTrackingObservers()
+        }
         self.updateSessionId(id: id)
-        self.sessionType == .automatic ? self.registerObserver() : self.deregisterObserver()
     }
```
[the remark this answers is not in the document]
Comment (SKannaniOS) [reply Sources/RudderStackAnalytics/SessionManagement/SessionHandler.swift:56]: When this class is initialized, it already contains the logic to register observers if automatic session tracking is enabled.

File: Sources/RudderStackAnalytics/Base/Analytics.swift [hunk 6]
```diff
@@ -316,11 +312,9 @@ extension Analytics {
         self.pluginChain?.removeAll()
         self.pluginChain = nil
         
-        self.lifecycleSessionWrapper?.invalidate()
-        self.lifecycleSessionWrapper = nil
-        
+        self.lifecycleObserver = nil
+        self.sessionHandler = nil
```
[the remark this answers is not in the document]
Comment (SKannaniOS) [reply Sources/RudderStackAnalytics/Base/Analytics.swift]: Since those are independent instances, `deinit` is enough.

File: Sources/RudderStackAnalytics/SessionManagement/SessionHandler.swift (hunk 5 above)
[the remark this answers is not in the document]
Comment (SKannaniOS) [reply Sources/RudderStackAnalytics/SessionManagement/SessionHandler.swift:56]: There is no public API to change session type at runtime.

File: Sources/RudderStackAnalytics/Base/Analytics.swift [hunk 7]
```diff
@@ -35,15 +35,14 @@ public class Analytics {
     private var processEventChannel: AsyncChannel<Event>
     
     /**
-     The handler instance responsible for managing lifecycle events and session-related operations.
-     
-     - Note: When this property is set, `startAutomaticSessionIfNeeded()` is automatically triggered to ensure the automatic session begins as required.
+     The observer responsible for monitoring application lifecycle events.
+    */
+    internal var lifecycleObserver: LifecycleObserver? = LifecycleObserver()
```
[the remark this answers is not in the document]
Comment (SKannaniOS) [reply Sources/RudderStackAnalytics/Base/Analytics.swift]: Yes. It is intentional.

File: Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift (hunk 3 above)
Comment (SKannaniOS) [reply Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift]: I have matched/refactored `SessionHandler` implementation along with related classes and test cases in this PR.

File: Sources/RudderStackAnalytics/Base/Analytics.swift (hunk 7 above)
Comment (1abhishekpandey) [inline Sources/RudderStackAnalytics/Base/Analytics.swift]: Should we initialise this in the `setup` method?

File: Sources/RudderStackAnalytics/LifecycleManagement/LifecycleManagementUtils.swift [hunk 8]
```diff
@@ -79,43 +79,12 @@ extension LifecycleEventListener {
  Wrapper to hold weak references to observers
  */
 class WeakObserver {
+    // Capture ObjectIdentifier instead of observer to avoid strong reference. So it won't extend the observer's lifetime.
+    let id: ObjectIdentifier
```
Comment (1abhishekpandey) [inline Sources/RudderStackAnalytics/LifecycleManagement/LifecycleManagementUtils.swift]: Let's rename the variable to something more meaningful. `id` doesn't seem a better variable name. How about `weakReference` or something else?

File: Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift [hunk 9]
```diff
@@ -63,10 +57,43 @@ extension LifecycleObserver {
...
         observers.append(WeakObserver(observer))
     }
     
-    func removeObserver(_ observer: LifecycleEventListener) {
-        observers.removeAll { $0.observer === observer }
+    func remove(byId id: ObjectIdentifier) {
+        observers.removeAll { $0.id == id }
+    }
+    
+    func snapshot() -> [LifecycleEventListener] {
```
Comment (1abhishekpandey) [inline Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift]: Can you explain why we need this method? As we shouldn't need a `snapshot` method in the observer class, `add` and `remove` should be the only two methods.

If we could remove this method somehow outside of the `LifecycleObserverStore`, let's do that. If that's not a good approach, let's add a comment.

File: Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift [hunk 10]
```diff
@@ -12,49 +12,43 @@ import Foundation
  A class created to observe app lifecycle events.
  */
 final class LifecycleObserver {
-    var analytics: Analytics?
-    
-    private var observers: [WeakObserver] = []
+    private let store = LifecycleObserverStore()
```
Comment (1abhishekpandey) [inline Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift]: Do you think a better variable name would be `lifecycleObserver` or `lifecycleObserverStore`? Current `store` variable name doesn't help in understanding the context.

File: Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift [hunk 11]
```diff
@@ -63,10 +57,43 @@ extension LifecycleObserver {
...
+            await store.add(observer)
+        }
+    }
+
+    func removeObserver(_ observer: LifecycleEventListener) {
+        // Capture ObjectIdentifier instead of observer to avoid strong reference.
+        // This is safe to call from deinit as it won't extend the observer's lifetime.
+        let observerId = ObjectIdentifier(observer)
+        Task { [store] in
+            await store.remove(byId: observerId)
```
Comment (1abhishekpandey) [inline Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift]: I just want to confirm this: While adding the `observer`, we don't wrap the observer in `ObjectIdentifier` (but while removing it, we do). 

1. Is it fine to not wrap it while adding the observer?
2. If we do not need it for adding, then why do we need it only while removing it?

(I'm trying to understand its importance over here.)

File: Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift [hunk 12]
```diff
@@ -12,49 +12,43 @@ import Foundation
...
-        self.analytics = analytics
-        self.registerNotifications()
+    init() {
+        registerNotifications()
     }
     
     deinit {
-        self.observers.removeAll()
-        self.notificationObservers.forEach { NotificationCenter.default.removeObserver($0) }
+        notificationObservers.forEach(NotificationCenter.default.removeObserver)
```
Comment (1abhishekpandey) [inline Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift:23]: Don't we need to clear the `store` variable?

**Review by 1abhishekpandey — COMMENTED** [review]
Great work

File: Sources/RudderStackAnalytics/LifecycleManagement/LifecycleManagementUtils.swift (hunk 8 above)
Comment (SKannaniOS) [reply Sources/RudderStackAnalytics/LifecycleManagement/LifecycleManagementUtils.swift]: That's an identifier. I'll rename to `identifier` of the `WeakObserver`.

File: Sources/RudderStackAnalytics/Base/Analytics.swift (hunk 7 above)
Comment (SKannaniOS) [reply Sources/RudderStackAnalytics/Base/Analytics.swift]: We can do that, but this class doesn’t need an analytics instance. So it can be initialized at declaration time. It doesn’t make any difference.

Comment (1abhishekpandey) [reply Sources/RudderStackAnalytics/Base/Analytics.swift]: In that case, let's define it in the `setup` method for consistency.

File: Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift (hunk 9 above)
Comment (SKannaniOS) [reply Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift]: This method returns a stable copy of active observers for safe iteration during event broadcasting, while also cleaning up any deallocated weak references. I'll add a function documentation.

File: Sources/RudderStackAnalytics/LifecycleManagement/LifecycleManagementUtils.swift (hunk 8 above)
Comment (1abhishekpandey) [reply Sources/RudderStackAnalytics/LifecycleManagement/LifecycleManagementUtils.swift]: If possible, try to add `WeakObserver` word in the variable name to signify its context.

File: Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift (hunk 10 above)
Comment (SKannaniOS) [reply Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift]: I'll rename to `observerStore`.

File: Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift (hunk 11 above)
Comment (SKannaniOS) [reply Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift]: For adding, we need the actual object to create a weak reference to it (captured weakly to avoid lifetime extension). For removing, we only need to identify which observer to remove —  `ObjectIdentifier` gives us a stable identifier that works even when called from deinit, where the object is already deallocating and a weak reference would be nil.

File: Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift (hunk 12 above)
Comment (SKannaniOS) [reply Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift:23]: No, we don't need to manually clear it. When `LifecycleObserver` is deallocated, the `LifecycleObserverStore` actor it owns will also be deallocated. Swift's ARC automatically cleans up the observers array along with all `WeakObserver` instances, because it holds already weak references only. I already tested it multiple times with `Shutdown` call.

File: Sources/RudderStackAnalytics/Base/Analytics.swift (hunk 7 above)
Comment (SKannaniOS) [reply Sources/RudderStackAnalytics/Base/Analytics.swift]: Ok.

File: Sources/RudderStackAnalytics/LifecycleManagement/LifecycleManagementUtils.swift (hunk 8 above)
Comment (SKannaniOS) [reply Sources/RudderStackAnalytics/LifecycleManagement/LifecycleManagementUtils.swift]: No. The class name itself `WeakObserver`, why we need to repeat it.?

Comment (1abhishekpandey) [reply Sources/RudderStackAnalytics/LifecycleManagement/LifecycleManagementUtils.swift]: We are discussing this offline.

Comment (SKannaniOS) [reply Sources/RudderStackAnalytics/LifecycleManagement/LifecycleManagementUtils.swift]: After the discussion, we aligned to go with simplified solution of `LifeCycleObserver`.

File: Sources/RudderStackAnalytics/Base/Analytics.swift [hunk 13]
```diff
@@ -369,8 +374,9 @@ extension Analytics {
         self.setupSourceConfig()
         self.startProcessingEvents()
         
+        self.lifecycleObserver = LifecycleObserver() // Should be initialized before SessionHandler
```
Comment (1abhishekpandey) [inline Sources/RudderStackAnalytics/Base/Analytics.swift:377]: 👍

**Review by 1abhishekpandey — APPROVED** [review]

File: Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift (hunk 10 above)
Comment (1abhishekpandey) [reply Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift]: I cannot find something called `observerStore`. Is this removed?

File: Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift (hunk 11 above)
Comment (1abhishekpandey) [reply Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift]: I cannot find `ObjectIdentifier`. Has this been removed?

File: Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift (hunk 10 above)
Comment (SKannaniOS) [reply Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift]: Yes. Removed.

File: Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift (hunk 11 above)
Comment (SKannaniOS) [reply Sources/RudderStackAnalytics/LifecycleManagement/LifecycleObserver.swift]: Yes. Removed.
